MAINT: Start testing on Python 3.11 in CI.

Carreau opened the pull request:
Python 3.11 was released in October 2022. 
In this PR, the CI test suite is extended to test on Python 3.11. 

I'm hoping that CI on Python 3.11 will be faster, as 3.11 is reported to be quite a bit faster in some use cases.

Additionally, there a few refactors and split of test into multiple subtests, as some of them are segfaulting on some platforms/python version combinations.

### Needs Tasks
- [x] https://github.com/scikit-image/scikit-image/issues/6517 (0.20 published)
- [x] https://github.com/google/tensorstore/issues/70
- [x] Vispy should get 3.11 wheel now that they've bump CI build wheel https://github.com/vispy/vispy/pull/2431

asciitree (has no wheels whatsoever).

 - [x] wrapt (currently 1.15.0 – 27 Feb),
 - [x] pytomlpp (via npe2),
 - [x] ~~MarkupSafe~~ also do not have binary artifacts, 
 - [x] lxml is an issue on osx-arm.

### Once merged:

Please open an issue to track the relevant PyQt5 failure that both seem to be related to `Label`, so there might actually be a bug there.

Carreau commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -154,15 +154,13 @@ jobs:
     steps:
       - uses: actions/checkout@v3
-        with:
-          path: napari-from-github
```
```suggestion
        with:
          path: napari-from-github
```

Carreau commented on .github/workflows/test_typing.yml:
```diff
@@ -12,12 +12,12 @@ jobs:
         run: |
           pip install -r resources/requirements_mypy.txt
-          SETUPTOOLS_ENABLE_FEATURES="legacy-editable" pip install -e .[all]
+          SETUPTOOLS_ENABLE_FEATURES="legacy-editable" pip install --only-binary ':all:' -e .[all]
```
I'm guessing this will fail as some deps don't have wheels.

Carreau commented on tox.ini:
```diff
@@ -51,6 +52,7 @@ BACKEND =
 # Settings defined in the top-level testenv section are automatically
 # inherited by individual environments unless overridden.
 [testenv]
+install_command = pip install --only-binary :all: {opts} {packages}
```
And this is likely to fail as well.

Carreau commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -177,7 +177,7 @@ jobs:
           python-version: 3.9
           cache: "pip"
-          cache-dependency-path: napari-from-github/setup.cfg
+          cache-dependency-path: setup.cfg
```
```suggestion
          cache-dependency-path: napari-from-github/setup.cfg
```

dstansby commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -93,10 +93,23 @@ jobs:
             platform: ubuntu-latest
             backend: pyside6
-
+          - python: '3.11'
```
Might be a bit clearer to use [`exclude`](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#excluding-matrix-configurations) here to exclude the Pyside2 3.11 build?

Carreau replied:
Yes, that's a good idea. Once the tests are passing I can try to do that.

Carreau committed:
MAINT: Try to bump Python versions in CI.

It seem that it's been some time since Python 3.11 is out, Try to use it
where relevant, and extend out test suite to test on Python 3.11. Where
relevant also use "3.x" which means latest stable, so it should
auto-bump to 3.12 when released, which avoids an extra bump.

I'm hopping this will slightly speed-up CI as Python 3.11 is reported to
be quite faster in some use case, but at least let us know if there are
things we are waiting for for for 3.11 compatibility.

Czaki commented on .github/workflows/test_comprehensive.yml:
```diff
@@ -171,7 +171,7 @@ jobs:
       - uses: actions/setup-python@v4
         with:
-          python-version: 3.9
+          python-version: "3.x"
```
As we run here `tox -e py39-linux-pyside2-examples`, then python 3.9 need to be installed

Carreau replied:
Sure.

```suggestion
          python-version: 3.9
```

Czaki commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -201,10 +208,13 @@ jobs:
         with:
-          python-version: 3.9
-          cache-dependency-path: setup.cfg
+          python-version: 3.11
```
As we run here `tox -e py39-linux-pyside2-examples`, then python 3.9 need to be installed or python version for check examples need to be changed

Carreau replied:
Should we rename to "test examples are working on older supported version" ?
```suggestion
          python-version: 3.9
```

And do we want to test examples on the newest python version ?

Czaki replied:
It was tested working on the middle version when implemented (as expected as more popular). So maybe it should be moved to python 3.10.

Czaki commented on resources/constraints/constraints_py3.11.txt:
```diff
@@ -0,0 +1,536 @@
+#
+# This file is autogenerated by pip-compile with Python 3.10
```
it should be done using python 3.11?

Carreau replied:
Yeah, sorry I could not get pip-compile to spit out something on Python 3.11 it hangs on my machine so, I copied the 3.10 file.

Czaki replied:
I have updated part of this PR connected with python 3.11 constraints.

Czaki reviewed (commented):
@tlambert03 Could you take a look at `py311-linux-pyqt5` test failure? As stacktrace ends on superqt call.

Czaki committed:
fix python 3.11 constraints

Carreau committed:
revert to hardcoded python 3.9

Carreau committed:
Bump doc building to 3.10.

SciPy 1.9.1 does not have a wheel, 1.10.1 has and thus consume why less
CI and does not fail

Carreau committed:
3.10 need to be quoted

Carreau commented on .github/workflows/upgrade_test_constraints.yml:
```diff
@@ -37,17 +37,25 @@ jobs:
           python3.9 -m piptools compile --upgrade -o resources/constraints/constraints_py3.9.txt setup.cfg resources/constraints/version_denylist.txt --extra pyqt5 --extra pyqt6_experimental --extra pyside2 --extra pyside6_experimental --extra testing --allow-unsafe --strip-extras --resolver=backtracking
           python3.9 -m piptools compile --upgrade -o resources/constraints/constraints_py3.9_examples.txt setup.cfg resources/constraints/version_denylist.txt resources/constraints/version_denylist_examples.txt --extra pyqt5 --extra pyqt6_experimental --extra pyside2 --extra pyside6_experimental --extra testing --allow-unsafe --strip-extras --resolver=backtracking
           python3.10 -m piptools compile --upgrade -o resources/constraints/constraints_py3.10.txt setup.cfg resources/constraints/version_denylist.txt --extra pyqt5 --extra pyqt6_experimental --extra pyside2 --extra pyside6_experimental --extra testing --allow-unsafe --strip-extras --resolver=backtracking
+          python3.11 -m piptools compile --upgrade -o resources/constraints/constraints_py3.11.txt setup.cfg resources/constraints/version_denylist.txt --extra pyqt5 --extra pyqt6_experimental --extra pyside2 --extra pyside6_experimental --extra testing --allow-unsafe --strip-extras --resolver=backtracking
```
Side question, do anyone belive it would be worth replacing this by something like

```
for pyv in 3.8 3.9 3.10 3.11; do
      python${pyv}  -m pip install -U pip pip-tools
      python${pyv}  -m piptools compile --upgrade -o resources/constraints/constraints_${pyv}.txt  setup.cfg resources/constraints/version_denylist.txt --extra pyqt5 --extra pyqt6_experimental --extra pyside2 --extra pyside6_experimental --extra testing --allow-unsafe --strip-extras --resolver=backtracking
done
```

And have the `py3.9_examples.txt` on its own line ?

Czaki replied:
I like it.

Carreau committed:
refactor compile file with loop

Carreau committed:
split test to narrow failure

Carreau committed:
Extract failing param

Czaki commented on .github/workflows/test_typing.yml:
```diff
@@ -12,7 +12,7 @@ jobs:
       - uses: actions/setup-python@v4
         with:
-          python-version: 3.9
+          python-version: "3.x"
```
mypy uses constrains files https://github.com/napari/napari/blob/2c5f3bfd453a3cf5dfc16c7da7e1dc6fe348f64e/tox.ini#L152-L157

So it need to have pinned python version.

Carreau replied:
Yeah, this PR stated in Dec 2022, the constraints files were started April 10 and merged 12 days later, they are leftovers.

Czaki replied:
I know. But as I remember, it takes a long time to have this PR working because of problems with dependencies. 

Also the general problem is that I expect that napari will not support python 3.12 shortly after release.

Czaki replied:
```suggestion
          python-version: "3.11"
```

Czaki reviewed (changes requested):
I understand why you want to use latest stable python instead of the pinned version in some workflows, but most often, napari needs to wait a few months until all dependencies release wheels for the new Python version. 

Especially napari has ipython in dependencies with known incompatibilities to python 3.12. So any workflow that requires installing napari should use python 3.11 instead of 3.x.

Czaki commented on .github/workflows/test_translations.yml:
```diff
@@ -12,10 +12,10 @@ jobs:
         uses: actions/setup-python@v4
         with:
-          python-version: 3.9
+          python-version: "3.x"
```
```suggestion
          python-version: "3.11"
```

Czaki commented on .github/workflows/test_translations.yml:
```diff
@@ -12,10 +12,10 @@ jobs:
     steps:
       - uses: actions/checkout@v3
-      - name: Set up Python 3.8
+      - name: Set up Python 3.x
```
```suggestion
      - name: Set up Python 3.11
```

Czaki committed:
Apply suggestions from code review

Czaki commented on .github/workflows/test_comprehensive.yml:
```diff
@@ -24,10 +24,10 @@ jobs:
     steps:
       - uses: actions/checkout@v3
-      - name: Set up Python 3.8
+      - name: "Set up Python 3.x (latest stable)"
```
```suggestion
      - name: "Set up Python 3.11"
```

Czaki commented on .github/workflows/test_comprehensive.yml:
```diff
@@ -24,10 +24,10 @@ jobs:
         uses: actions/setup-python@v4
         with:
-          python-version: 3.8
+          python-version: "3.x"
```
```suggestion
          python-version: "3.11"
```

Czaki reviewed (commented):
for consistency

Czaki committed:
Apply suggestions from code review
